File: 08_pairSwapNodes/8.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* createNode(int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

struct Node* createLinkedList(int values[], int size) {
    if (size == 0)
        return NULL;

    struct Node* head = createNode(values[0]);
    struct Node* current = head;

    for (int i = 1; i < size; i++) {
        current->next = createNode(values[i]);
        current = current->next;
    }

    return head;
}

void printLinkedList(struct Node* head) {
    struct Node* current = head;
    while (current) {
        printf("%d ", current->data);
        current = current->next;
    }
    printf("\n");
}

struct Node* pairwiseSwap(struct Node* head) {
    struct Node* first = head;
    struct Node* second = head ? head->next : NULL;

    while (first && second) {
        int temp = first->data;
        first->data = second->data;
        second->data = temp;

        first = second->next;
        if (first)
            second = first->next;
        else
            break;
    }

    return head;
}

int main() {
    int values[] = {1, 2, 3, 4, 5, 6};
    int size = sizeof(values) / sizeof(values[0]);

    struct Node* head = createLinkedList(values, size);

    printf("Original list:\n");
    printLinkedList(head);

    head = pairwiseSwap(head);

    printf("After pairwise swap:\n");
    printLinkedList(head);

    return 0;
}

File: 08_pairSwapNodes/8.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Solution:
    def pairwiseSwap(self, head):
        first = head
        second = head.next

        while first and second:
            temp = first.data
            first.data = second.data
            second.data = temp

            first = second.next
            if first:
                second = first.next
            else:
                break

        return head

def create_linked_list(values):
    if not values:
        return None
    head = Node(values[0])
    current = head
    for val in values[1:]:
        current.next = Node(val)
        current = current.next
    return head

def print_linked_list(head):
    current = head
    while current:
        print(current.data, end=" ")
        current = current.next
    print()

values = [1, 2, 3, 4, 5, 6]
head = create_linked_list(values)

print("Original list:")
print_linked_list(head)

sol = Solution()
swapped_head = sol.pairwiseSwap(head)

print("After pairwise swap:")
print_linked_list(swapped_head)